Zenfs dev 1.4

levichen94 opened the pull request:
WIP: merge fs abstraction only

mrambacher committed:
The ObjectRegistry class replaces the Registrar and NewCustomObjects.… (#5293)

Summary:
The ObjectRegistry class replaces the Registrar and NewCustomObjects.  Objects are registered with the registry by Type (the class must implement the static const char *Type() method).

This change is necessary for a few reasons:
- By having a class (rather than static template instances), the class can be passed between compilation units, meaning that objects could be registered and shared from a dynamic library with an executable.
- By having a class with instances, different units could have different objects registered.  This could be useful if, for example, one Option allowed for a dynamic library and one did not.

When combined with some other PRs (being able to load shared libraries, a Configurable interface to configure objects to/from string), this code will allow objects in external shared libraries to be added to a RocksDB image at run-time, rather than requiring every new extension to be built into the main library and called explicitly by every program.

Test plan (on riversand963's  devserver)
```
$COMPILE_WITH_ASAN=1 make -j32 all && sleep 1 && make check
```
All tests pass.
Pull Request resolved: https://github.com/facebook/rocksdb/pull/5293

Differential Revision: D16363396

Pulled By: riversand963

fbshipit-source-id: fbe4acb615bfc11103eef40a0b288845791c0180
Signed-off-by: Changlong Chen <levisonchen@live.cn>

Commit:
Introduce a new storage specific Env API (#5761)

Summary:
The current Env API encompasses both storage/file operations, as well as OS related operations. Most of the APIs return a Status, which does not have enough metadata about an error, such as whether its retry-able or not, scope (i.e fault domain) of the error etc., that may be required in order to properly handle a storage error. The file APIs also do not provide enough control over the IO SLA, such as timeout, prioritization, hinting about placement and redundancy etc.

This PR separates out the file/storage APIs from Env into a new FileSystem class. The APIs are updated to return an IOStatus with metadata about the error, as well as to take an IOOptions structure as input in order to allow more control over the IO.

The user can set both ```options.env``` and ```options.file_system``` to specify that RocksDB should use the former for OS related operations and the latter for storage operations. Internally, a ```CompositeEnvWrapper``` has been introduced that inherits from ```Env``` and redirects individual methods to either an ```Env``` implementation or the ```FileSystem``` as appropriate. When options are sanitized during ```DB::Open```, ```options.env``` is replaced with a newly allocated ```CompositeEnvWrapper``` instance if both env and file_system have been specified. This way, the rest of the RocksDB code can continue to function as before.

This PR also ports PosixEnv to the new API by splitting it into two - PosixEnv and PosixFileSystem. PosixEnv is defined as a sub-class of CompositeEnvWrapper, and threading/time functions are overridden with Posix specific implementations in order to avoid an extra level of indirection.

The ```CompositeEnvWrapper``` translates ```IOStatus``` return code to ```Status```, and sets the severity to ```kSoftError``` if the io_status is retryable. The error handling code in RocksDB can then recover the DB automatically.
Pull Request resolved: https://github.com/facebook/rocksdb/pull/5761

Differential Revision: D18868376

Pulled By: anand1976

fbshipit-source-id: 39efe18a162ea746fabac6360ff529baba48486f
Signed-off-by: Changlong Chen <levisonchen@live.cn>

Commit:
Introduce a new storage specific Env API (#5761)

Summary:
The current Env API encompasses both storage/file operations, as well as OS related operations. Most of the APIs return a Status, which does not have enough metadata about an error, such as whether its retry-able or not, scope (i.e fault domain) of the error etc., that may be required in order to properly handle a storage error. The file APIs also do not provide enough control over the IO SLA, such as timeout, prioritization, hinting about placement and redundancy etc.

This PR separates out the file/storage APIs from Env into a new FileSystem class. The APIs are updated to return an IOStatus with metadata about the error, as well as to take an IOOptions structure as input in order to allow more control over the IO.

The user can set both ```options.env``` and ```options.file_system``` to specify that RocksDB should use the former for OS related operations and the latter for storage operations. Internally, a ```CompositeEnvWrapper``` has been introduced that inherits from ```Env``` and redirects individual methods to either an ```Env``` implementation or the ```FileSystem``` as appropriate. When options are sanitized during ```DB::Open```, ```options.env``` is replaced with a newly allocated ```CompositeEnvWrapper``` instance if both env and file_system have been specified. This way, the rest of the RocksDB code can continue to function as before.

This PR also ports PosixEnv to the new API by splitting it into two - PosixEnv and PosixFileSystem. PosixEnv is defined as a sub-class of CompositeEnvWrapper, and threading/time functions are overridden with Posix specific implementations in order to avoid an extra level of indirection.

The ```CompositeEnvWrapper``` translates ```IOStatus``` return code to ```Status```, and sets the severity to ```kSoftError``` if the io_status is retryable. The error handling code in RocksDB can then recover the DB automatically.
Pull Request resolved: https://github.com/facebook/rocksdb/pull/5761

Differential Revision: D18868376

Pulled By: anand1976

fbshipit-source-id: 39efe18a162ea746fabac6360ff529baba48486f
Signed-off-by: Changlong Chen <levisonchen@live.cn>

Commit:
Introduce a new storage specific Env API (#5761)

Summary:
The current Env API encompasses both storage/file operations, as well as OS related operations. Most of the APIs return a Status, which does not have enough metadata about an error, such as whether its retry-able or not, scope (i.e fault domain) of the error etc., that may be required in order to properly handle a storage error. The file APIs also do not provide enough control over the IO SLA, such as timeout, prioritization, hinting about placement and redundancy etc.

This PR separates out the file/storage APIs from Env into a new FileSystem class. The APIs are updated to return an IOStatus with metadata about the error, as well as to take an IOOptions structure as input in order to allow more control over the IO.

The user can set both ```options.env``` and ```options.file_system``` to specify that RocksDB should use the former for OS related operations and the latter for storage operations. Internally, a ```CompositeEnvWrapper``` has been introduced that inherits from ```Env``` and redirects individual methods to either an ```Env``` implementation or the ```FileSystem``` as appropriate. When options are sanitized during ```DB::Open```, ```options.env``` is replaced with a newly allocated ```CompositeEnvWrapper``` instance if both env and file_system have been specified. This way, the rest of the RocksDB code can continue to function as before.

This PR also ports PosixEnv to the new API by splitting it into two - PosixEnv and PosixFileSystem. PosixEnv is defined as a sub-class of CompositeEnvWrapper, and threading/time functions are overridden with Posix specific implementations in order to avoid an extra level of indirection.

The ```CompositeEnvWrapper``` translates ```IOStatus``` return code to ```Status```, and sets the severity to ```kSoftError``` if the io_status is retryable. The error handling code in RocksDB can then recover the DB automatically.
Pull Request resolved: https://github.com/facebook/rocksdb/pull/5761

Differential Revision: D18868376

Pulled By: anand1976

fbshipit-source-id: 39efe18a162ea746fabac6360ff529baba48486f
Signed-off-by: Changlong Chen <levisonchen@live.cn>

siying committed:
Replace namespace name "rocksdb" with ROCKSDB_NAMESPACE (#6433)

Summary:
When dynamically linking two binaries together, different builds of RocksDB from two sources might cause errors. To provide a tool for user to solve the problem, the RocksDB namespace is changed to a flag which can be overridden in build time.
Pull Request resolved: https://github.com/facebook/rocksdb/pull/6433

Test Plan: Build release, all and jtest. Try to build with ROCKSDB_NAMESPACE with another flag.

Differential Revision: D19977691

fbshipit-source-id: aa7f2d0972e1c31d75339ac48478f34f6cfcfb3e
Signed-off-by: Changlong Chen <levisonchen@live.cn>

siying committed:
Replace namespace name "rocksdb" with ROCKSDB_NAMESPACE (#6433)

Summary:
When dynamically linking two binaries together, different builds of RocksDB from two sources might cause errors. To provide a tool for user to solve the problem, the RocksDB namespace is changed to a flag which can be overridden in build time.
Pull Request resolved: https://github.com/facebook/rocksdb/pull/6433

Test Plan: Build release, all and jtest. Try to build with ROCKSDB_NAMESPACE with another flag.

Differential Revision: D19977691

fbshipit-source-id: aa7f2d0972e1c31d75339ac48478f34f6cfcfb3e
Signed-off-by: Changlong Chen <levisonchen@live.cn>

siying committed:
Replace namespace name "rocksdb" with ROCKSDB_NAMESPACE (#6433)

Summary:
When dynamically linking two binaries together, different builds of RocksDB from two sources might cause errors. To provide a tool for user to solve the problem, the RocksDB namespace is changed to a flag which can be overridden in build time.
Pull Request resolved: https://github.com/facebook/rocksdb/pull/6433

Test Plan: Build release, all and jtest. Try to build with ROCKSDB_NAMESPACE with another flag.

Differential Revision: D19977691

fbshipit-source-id: aa7f2d0972e1c31d75339ac48478f34f6cfcfb3e
Signed-off-by: Changlong Chen <levisonchen@live.cn>

siying committed:
Replace namespace name "rocksdb" with ROCKSDB_NAMESPACE (#6433)

Summary:
When dynamically linking two binaries together, different builds of RocksDB from two sources might cause errors. To provide a tool for user to solve the problem, the RocksDB namespace is changed to a flag which can be overridden in build time.
Pull Request resolved: https://github.com/facebook/rocksdb/pull/6433

Test Plan: Build release, all and jtest. Try to build with ROCKSDB_NAMESPACE with another flag.

Differential Revision: D19977691

fbshipit-source-id: aa7f2d0972e1c31d75339ac48478f34f6cfcfb3e
Signed-off-by: Changlong Chen <levisonchen@live.cn>

zhichao-cao committed:
Introduce FaultInjectionTestFS to test fault File system instead of Env (#6414)

Summary:
In the current code base, we can use FaultInjectionTestEnv to simulate the env issue such as file write/read errors, which are used in most of the test. The PR https://github.com/facebook/rocksdb/issues/5761 introduce the File System as a new Env API. This PR implement the FaultInjectionTestFS, which can be used to simulate when File System has issues such as IO error. user can specify any IOStatus error as input, such that FS corresponding actions will return certain error to the caller.

A set of ErrorHandlerFSTests are introduced for testing
Pull Request resolved: https://github.com/facebook/rocksdb/pull/6414

Test Plan: pass make asan_check, pass error_handler_fs_test.

Differential Revision: D20252421

Pulled By: zhichao-cao

fbshipit-source-id: e922038f8ce7e6d1da329fd0bba7283c4b779a21
Signed-off-by: Changlong Chen <levisonchen@live.cn>

Commit:
Cache result of GetLogicalBufferSize in Linux (#6457)

Summary:
In Linux, when reopening DB with many SST files, profiling shows that 100% system cpu time spent for a couple of seconds for `GetLogicalBufferSize`. This slows down MyRocks' recovery time when site is down.

This PR introduces two new APIs:
1. `Env::RegisterDbPaths` and `Env::UnregisterDbPaths` lets `DB` tell the env when it starts or stops using its database directories . The `PosixFileSystem` takes this opportunity to set up a cache from database directories to the corresponding logical block sizes.
2. `LogicalBlockSizeCache` is defined only for OS_LINUX to cache the logical block sizes.

Other modifications:
1. rename `logical buffer size` to `logical block size` to be consistent with Linux terms.
2. declare `GetLogicalBlockSize` in `PosixHelper` to expose it to `PosixFileSystem`.
3. change the functions `IOError` and `IOStatus` in `env/io_posix.h` to have external linkage since they are used in other translation units too.
Pull Request resolved: https://github.com/facebook/rocksdb/pull/6457

Test Plan:
1. A new unit test is added for `LogicalBlockSizeCache` in `env/io_posix_test.cc`.
2. A new integration test is added for `DB` operations related to the cache in `db/db_logical_block_size_cache_test.cc`.

`make check`

Differential Revision: D20131243

Pulled By: cheng-chang

fbshipit-source-id: 3077c50f8065c0bffb544d8f49fb10bba9408d04
Signed-off-by: Changlong Chen <levisonchen@live.cn>

Commit:
Simplify migration to FileSystem API (#6552)

Summary:
The current Env/FileSystem API separation has a couple of issues -
1. It requires the user to specify 2 options - ```Options::env``` and ```Options::file_system``` - which means they have to make code changes to benefit from the new APIs. Furthermore, there is a risk of accessing the same APIs in two different ways, through Env in the old way and through FileSystem in the new way. The two may not always match, for example, if env is ```PosixEnv``` and FileSystem is a custom implementation. Any stray RocksDB calls to env will use the ```PosixEnv``` implementation rather than the file_system implementation.
2. There needs to be a simple way for the FileSystem developer to instantiate an Env for backward compatibility purposes.

This PR solves the above issues and simplifies the migration in the following ways -
1. Embed a shared_ptr to the ```FileSystem``` in the ```Env```, and remove ```Options::file_system``` as a configurable option. This way, no code changes will be required in application code to benefit from the new API. The default Env constructor uses a ```LegacyFileSystemWrapper``` as the embedded ```FileSystem```.
1a. - This also makes it more robust by ensuring that even if RocksDB
  has some stray calls to Env APIs rather than FileSystem, they will go
  through the same object and thus there is no risk of getting out of
  sync.
2. Provide a ```NewCompositeEnv()``` API that can be used to construct a
PosixEnv with a custom FileSystem implementation. This eliminates an
indirection to call Env APIs, and relieves the FileSystem developer of
the burden of having to implement wrappers for the Env APIs.
3. Add a couple of missing FileSystem APIs - ```SanitizeEnvOptions()``` and
```NewLogger()```

Tests:
1. New unit tests
2. make check and make asan_check
Pull Request resolved: https://github.com/facebook/rocksdb/pull/6552

Reviewed By: riversand963

Differential Revision: D20592038

Pulled By: anand1976

fbshipit-source-id: c3801ad4153f96d21d5a3ae26c92ba454d1bf1f7
Signed-off-by: Changlong Chen <levisonchen@live.cn>

Commit:
Simplify migration to FileSystem API (#6552)

Summary:
The current Env/FileSystem API separation has a couple of issues -
1. It requires the user to specify 2 options - ```Options::env``` and ```Options::file_system``` - which means they have to make code changes to benefit from the new APIs. Furthermore, there is a risk of accessing the same APIs in two different ways, through Env in the old way and through FileSystem in the new way. The two may not always match, for example, if env is ```PosixEnv``` and FileSystem is a custom implementation. Any stray RocksDB calls to env will use the ```PosixEnv``` implementation rather than the file_system implementation.
2. There needs to be a simple way for the FileSystem developer to instantiate an Env for backward compatibility purposes.

This PR solves the above issues and simplifies the migration in the following ways -
1. Embed a shared_ptr to the ```FileSystem``` in the ```Env```, and remove ```Options::file_system``` as a configurable option. This way, no code changes will be required in application code to benefit from the new API. The default Env constructor uses a ```LegacyFileSystemWrapper``` as the embedded ```FileSystem```.
1a. - This also makes it more robust by ensuring that even if RocksDB
  has some stray calls to Env APIs rather than FileSystem, they will go
  through the same object and thus there is no risk of getting out of
  sync.
2. Provide a ```NewCompositeEnv()``` API that can be used to construct a
PosixEnv with a custom FileSystem implementation. This eliminates an
indirection to call Env APIs, and relieves the FileSystem developer of
the burden of having to implement wrappers for the Env APIs.
3. Add a couple of missing FileSystem APIs - ```SanitizeEnvOptions()``` and
```NewLogger()```

Tests:
1. New unit tests
2. make check and make asan_check
Pull Request resolved: https://github.com/facebook/rocksdb/pull/6552

Reviewed By: riversand963

Differential Revision: D20592038

Pulled By: anand1976

fbshipit-source-id: c3801ad4153f96d21d5a3ae26c92ba454d1bf1f7
Signed-off-by: Changlong Chen <levisonchen@live.cn>

zhichao-cao committed:
Pass IOStatus to write path and set retryable IO Error as hard error in BG jobs (#6487)

Summary:
In the current code base, we use Status to get and store the returned status from the call. Specifically, for IO related functions, the current Status cannot reflect the IO Error details such as error scope, error retryable attribute, and others. With the implementation of https://github.com/facebook/rocksdb/issues/5761, we have the new Wrapper for IO, which returns IOStatus instead of Status. However, the IOStatus is purged at the lower level of write path and transferred to Status.

The first job of this PR is to pass the IOStatus to the write path (flush, WAL write, and Compaction). The second job is to identify the Retryable IO Error as HardError, and set the bg_error_ as HardError. In this case, the DB Instance becomes read only. User is informed of the Status and need to take actions to deal with it (e.g., call db->Resume()).
Pull Request resolved: https://github.com/facebook/rocksdb/pull/6487

Test Plan: Added the testing case to error_handler_fs_test. Pass make asan_check

Reviewed By: anand1976

Differential Revision: D20685017

Pulled By: zhichao-cao

fbshipit-source-id: ff85f042896243abcd6ef37877834e26f36b6eb0
Signed-off-by: Changlong Chen <levisonchen@live.cn>

zhichao-cao committed:
Pass IOStatus to write path and set retryable IO Error as hard error in BG jobs (#6487)

Summary:
In the current code base, we use Status to get and store the returned status from the call. Specifically, for IO related functions, the current Status cannot reflect the IO Error details such as error scope, error retryable attribute, and others. With the implementation of https://github.com/facebook/rocksdb/issues/5761, we have the new Wrapper for IO, which returns IOStatus instead of Status. However, the IOStatus is purged at the lower level of write path and transferred to Status.

The first job of this PR is to pass the IOStatus to the write path (flush, WAL write, and Compaction). The second job is to identify the Retryable IO Error as HardError, and set the bg_error_ as HardError. In this case, the DB Instance becomes read only. User is informed of the Status and need to take actions to deal with it (e.g., call db->Resume()).
Pull Request resolved: https://github.com/facebook/rocksdb/pull/6487

Test Plan: Added the testing case to error_handler_fs_test. Pass make asan_check

Reviewed By: anand1976

Differential Revision: D20685017

Pulled By: zhichao-cao

fbshipit-source-id: ff85f042896243abcd6ef37877834e26f36b6eb0
Signed-off-by: Changlong Chen <levisonchen@live.cn>

zhichao-cao committed:
Pass IOStatus to write path and set retryable IO Error as hard error in BG jobs (#6487)

Summary:
In the current code base, we use Status to get and store the returned status from the call. Specifically, for IO related functions, the current Status cannot reflect the IO Error details such as error scope, error retryable attribute, and others. With the implementation of https://github.com/facebook/rocksdb/issues/5761, we have the new Wrapper for IO, which returns IOStatus instead of Status. However, the IOStatus is purged at the lower level of write path and transferred to Status.

The first job of this PR is to pass the IOStatus to the write path (flush, WAL write, and Compaction). The second job is to identify the Retryable IO Error as HardError, and set the bg_error_ as HardError. In this case, the DB Instance becomes read only. User is informed of the Status and need to take actions to deal with it (e.g., call db->Resume()).
Pull Request resolved: https://github.com/facebook/rocksdb/pull/6487

Test Plan: Added the testing case to error_handler_fs_test. Pass make asan_check

Reviewed By: anand1976

Differential Revision: D20685017

Pulled By: zhichao-cao

fbshipit-source-id: ff85f042896243abcd6ef37877834e26f36b6eb0
Signed-off-by: Changlong Chen <levisonchen@live.cn>

riversand963 committed:
Add IsDirectory() to Env and FS (#6711)

Summary:
IsDirectory() is a common API to check whether a path is a regular file or
directory.
POSIX: call stat() and use S_ISDIR(st_mode)
Windows: PathIsDirectoryA() and PathIsDirectoryW()
HDFS: FileSystem.IsDirectory()
Java: File.IsDirectory()
...
Pull Request resolved: https://github.com/facebook/rocksdb/pull/6711

Test Plan: make check

Reviewed By: anand1976

Differential Revision: D21053520

Pulled By: riversand963

fbshipit-source-id: 680aadfd8ce982b63689190cf31b3145d5a89e27
Signed-off-by: Changlong Chen <levisonchen@live.cn>

pdillinger committed:
C++20 compatibility (#6697)

Summary:
Based on https://github.com/facebook/rocksdb/issues/6648 (CLA Signed), but heavily modified / extended:

* Implicit capture of this via [=] deprecated in C++20, and [=,this] not standard before C++20 -> now using explicit capture lists
* Implicit copy operator deprecated in gcc 9 -> add explicit '= default' definition
* std::random_shuffle deprecated in C++17 and removed in C++20 -> migrated to a replacement in RocksDB random.h API
* Add the ability to build with different std version though -DCMAKE_CXX_STANDARD=11/14/17/20 on the cmake command line
* Minimal rebuild flag of MSVC is deprecated and is forbidden with /std:c++latest (C++20)
* Added MSVC 2019 C++11 & MSVC 2019 C++20 in AppVeyor
* Added GCC 9 C++11 & GCC9 C++20 in Travis
Pull Request resolved: https://github.com/facebook/rocksdb/pull/6697

Test Plan: make check and CI

Reviewed By: cheng-chang

Differential Revision: D21020318

Pulled By: pdillinger

fbshipit-source-id: 12311be5dbd8675a0e2c817f7ec50fa11c18ab91
Signed-off-by: Changlong Chen <levisonchen@live.cn>

Commit:
Pass a timeout to FileSystem for random reads (#6751)

Summary:
Calculate ```IOOptions::timeout``` using ```ReadOptions::deadline``` and pass it to ```FileSystem::Read/FileSystem::MultiRead```. This allows us to impose a tighter bound on the time taken by Get/MultiGet on FileSystem/Envs that support IO timeouts. Even on those that don't support, check in ```RandomAccessFileReader::Read``` and ```MultiRead``` and return ```Status::TimedOut()``` if the deadline is exceeded.

For now, TableReader creation, which might do file opens and reads, are not covered. It will be implemented in another PR.

Tests:
Update existing unit tests to verify the correct timeout value is being passed
Pull Request resolved: https://github.com/facebook/rocksdb/pull/6751

Reviewed By: riversand963

Differential Revision: D21285631

Pulled By: anand1976

fbshipit-source-id: d89af843e5a91ece866e87aa29438b52a65a8567
Signed-off-by: Changlong Chen <levisonchen@live.cn>

ajkr committed:
prototype status check enforcement (#6798)

Summary:
Tried making Status object enforce that it is checked in some way. In cases it is not checked, `PermitUncheckedError()` must be called explicitly.

Added a way to run tests (`ASSERT_STATUS_CHECKED=1 make -j48 check`) on a
whitelist. The effort appears significant to get each test to pass with
this assertion, so I only fixed up enough to get one test (`options_test`)
working and added it to the whitelist.
Pull Request resolved: https://github.com/facebook/rocksdb/pull/6798

Reviewed By: pdillinger

Differential Revision: D21377404

Pulled By: ajkr

fbshipit-source-id: 73236f9c8df38f01cf24ecac4a6d1661b72d077e
Signed-off-by: Changlong Chen <levisonchen@live.cn>

yhr committed:
[PATCH 3/4] Add zenfs testing and benchmarking scripts

TerarkDB supports Zone Namespace SSD transplantation of
ZenFS patch v3, direct git am without any adaptation.

The following is the original patch description of Hans:
---------------------------------------------------------
Add a bunch of testing and benchmarking scripts.
Hopefully self-documenting :)

Signed-off-by: Yuanliang Wang <wangyuanliang@bytedance.com>

Commit:
Add FileSystem wrapper classes for IO tracing. (#7002)

Summary:
1. Add the wrapper classes FileSystemTracingWrapper, FSSequentialFileTracingWrapper, FSRandomAccessFileTracingWrapper, FSWritableFileTracingWrapper, FSRandomRWFileTracingWrapper that forward the calls to underlying storage system and then pass the file operation information to IOTracer. IOTracer dumps the record in binary format for tracing.
2. Add the wrapper classes FileSystemPtr, FSSequentialFilePtr, FSRandomAccessFilePtr, FSWritableFilePtr and FSRandomRWFilePtr that overload operator-> and return ptr to underlying storage system or Tracing wrapper class based on enabling/disabling of IO tracing. These classes are added to bypass Tracing Wrapper classes when we disable tracing.
3. Add enums in trace.h that distinguish which options need to be added for different file operations(Read, close, write etc) as part of tracing record.

Pull Request resolved: https://github.com/facebook/rocksdb/pull/7002

Test Plan: make check -j64

Reviewed By: anand1976

Differential Revision: D22127897

Pulled By: akankshamahajan15

fbshipit-source-id: 74cff58ce5661c9a3832dfaa52483f3b2d8565e0
Signed-off-by: Changlong Chen <levisonchen@live.cn>

mrambacher committed:
Make env*_test work with ASSERT_STATUS_CHECKED (#7176)

Summary:
Make (most of) the env*_test pass when ASSERT_STATUS_CHECKED is enabled.

One test that opens a database is currently disabled in this mode, as there are many errors that need revisited for DB tests and status checks.

Pull Request resolved: https://github.com/facebook/rocksdb/pull/7176

Reviewed By: cheng-chang

Differential Revision: D22799278

Pulled By: ajkr

fbshipit-source-id: 16d8a02eaeecd6df1060249b6a5811292801f2ed
Signed-off-by: Changlong Chen <levisonchen@live.cn>

acelyc111 committed:
Remove duplicate colon in Status message (#7041)

Summary:
A colon will be added after 'msg' automatically when invoke function Status(Code _code, const Slice& msg, const Slice& msg2),
it's not needed to append a colon explicitly to 'msg'.
Pull Request resolved: https://github.com/facebook/rocksdb/pull/7041

Reviewed By: ajkr

Differential Revision: D22292801

fbshipit-source-id: 8f2d69065bb779d2613468bf9fc9169f32c3f1ec
Signed-off-by: Changlong Chen <levisonchen@live.cn>

jay-zhuang committed:
Add buffer prefetch support for non directIO usecase (#7312)

Summary:
A new file interface `SupportPrefetch()` is added. When the user overrides it to `false`, an internal prefetch buffer will be used for readahead. Useful for non-directIO but FS doesn't have readahead support.

Pull Request resolved: https://github.com/facebook/rocksdb/pull/7312

Reviewed By: anand1976

Differential Revision: D23329847

Pulled By: jay-zhuang

fbshipit-source-id: 71cd4ce6f4a820840294e4e6aec111ab76175527
Signed-off-by: Changlong Chen <levisonchen@live.cn>

Commit:
Add a new IOStatus subcode to indicate that writes are fenced off (#7374)

Summary:
In a distributed file system, directory ownership is enforced by fencing
off the previous owner once they've been preempted by a new owner. This
PR adds a IOStatus subcode for ```StatusCode::IOError``` to indicate this.
Once this error is returned for a file write, the DB is put in read-only
mode and not allowed to resume in read-write mode.

Pull Request resolved: https://github.com/facebook/rocksdb/pull/7374

Test Plan: Add new unit tests in ```error_handler_fs_test```

Reviewed By: riversand963

Differential Revision: D23687777

Pulled By: anand1976

fbshipit-source-id: bef948642089dc0af399057864d9a8ca339e8b2f
Signed-off-by: Changlong Chen <levisonchen@live.cn>

zhichao-cao committed:
Add AppendWithVerify and PositionedAppendWithVerify to Env and FileSystem (#7419)

Summary:
Add new AppendWithVerify and PositionedAppendWithVerify APIs to Env and FileSystem to bring the data verification information (data checksum information) from upper layer (e.g., WritableFileWriter) to the storage layer. This PR only include the API definition, no functional codes are added to unblock other developers which depend on these APIs.

Pull Request resolved: https://github.com/facebook/rocksdb/pull/7419

Test Plan: make -j32

Reviewed By: pdillinger

Differential Revision: D23883196

Pulled By: zhichao-cao

fbshipit-source-id: 94676c26bc56144cc32e3661f84f21eccd790411
Signed-off-by: Changlong Chen <levisonchen@live.cn>

mrambacher committed:
Fix many tests to run with MEM_ENV and ENCRYPTED_ENV; Introduce a MemoryFileSystem class (#7566)

Summary:
This PR does a few things:

1.  The MockFileSystem class was split out from the MockEnv.  This change would theoretically allow a MockFileSystem to be used by other Environments as well (if we created a means of constructing one).  The MockFileSystem implements a FileSystem in its entirety and does not rely on any Wrapper implementation.

2.  Make the RocksDB test suite work when MOCK_ENV=1 and ENCRYPTED_ENV=1 are set.  To accomplish this, a few things were needed:
- The tests that tried to use the "wrong" environment (Env::Default() instead of env_) were updated
- The MockFileSystem was changed to support the features it was missing or mishandled (such as recursively deleting files in a directory or supporting renaming of a directory).

3.  Updated the test framework to have a ROCKSDB_GTEST_SKIP macro.  This can be used to flag tests that are skipped.  Currently, this defaults to doing nothing (marks the test as SUCCESS) but will mark the tests as SKIPPED when RocksDB is upgraded to a version of gtest that supports this (gtest-1.10).

I have run a full "make check" with MEM_ENV, ENCRYPTED_ENV,  both, and neither under both MacOS and RedHat.  A few tests were disabled/skipped for the MEM/ENCRYPTED cases.  The error_handler_fs_test fails/hangs for MEM_ENV (presumably a timing problem) and I will introduce another PR/issue to track that problem.  (I will also push a change to disable those tests soon).  There is one more test in DBTest2 that also fails which I need to investigate or skip before this PR is merged.

Theoretically, this PR should also allow the test suite to run against an Env loaded from the registry, though I do not have one to try it with currently.

Finally, once this is accepted, it would be nice if there was a CircleCI job to run these tests on a checkin so this effort does not become stale.  I do not know how to do that, so if someone could write that job, it would be appreciated :)

Pull Request resolved: https://github.com/facebook/rocksdb/pull/7566

Reviewed By: zhichao-cao

Differential Revision: D24408980

Pulled By: jay-zhuang

fbshipit-source-id: 911b1554a4d0da06fd51feca0c090a4abdcb4a5f
Signed-off-by: Changlong Chen <levisonchen@live.cn>

mm304321141 committed:
[refactor] Auto fallback mmap reads

royguo committed:
[docs] Update README

mrambacher committed:
Create a CustomEnv class; Add WinFileSystem; Make LegacyFileSystemWrapper private (#7703)

Summary:
This PR does the following:
-> Creates a WinFileSystem class.  This class is the Windows equivalent of the PosixFileSystem and will be used on Windows systems.
-> Introduces a CustomEnv class.  A CustomEnv is an Env that takes a FileSystem as constructor argument.  I believe there will only ever be two implementations of this class (PosixEnv and WinEnv).  There is still a CustomEnvWrapper class that takes an Env and a FileSystem and wraps the Env calls with the input Env but uses the FileSystem for the FileSystem calls
-> Eliminates the public uses of the LegacyFileSystemWrapper.

With this change in place, there are effectively the following patterns of Env:
- "Base Env classes" (PosixEnv, WinEnv).  These classes implement the core Env functions (e.g. Threads) and have a hard-coded input FileSystem.  These classes inherit from CompositeEnv, implement the core Env functions (threads) and delegate the FileSystem-like calls to the input file system.
- Wrapped Composite Env classes (MemEnv).  These classes take in an Env and a FileSystem.  The core env functions are re-directed to the wrapped env.  The file system calls are redirected to the input file system
- Legacy Wrapped Env classes.  These classes take in an Env input (but no FileSystem).  The core env functions are re-directed to the wrapped env.  A "Legacy File System" is created using this env and the file system calls directed to the env itself.

With these changes in place, the PosixEnv becomes a singleton -- there is only ever one created.  Any other use of the PosixEnv is via another wrapped env.  This cleans up some of the issues with the env construction and destruction.

Additionally, there were places in the code that required had an Env when they required a FileSystem.  Many of these places would wrap the Env with a LegacyFileSystemWrapper instead of using the env->GetFileSystem().  These places were changed, thereby removing layers of additional redirection (LegacyFileSystem --> Env --> Env::FileSystem).

Pull Request resolved: https://github.com/facebook/rocksdb/pull/7703

Reviewed By: zhichao-cao

Differential Revision: D25762190

Pulled By: anand1976

fbshipit-source-id: 1a088e97fc916f28ac69c149cd1dcad0ab31704b
Signed-off-by: Changlong Chen <levisonchen@live.cn>

mrambacher committed:
Create a CustomEnv class; Add WinFileSystem; Make LegacyFileSystemWrapper private (#7703)

Summary:
This PR does the following:
-> Creates a WinFileSystem class.  This class is the Windows equivalent of the PosixFileSystem and will be used on Windows systems.
-> Introduces a CustomEnv class.  A CustomEnv is an Env that takes a FileSystem as constructor argument.  I believe there will only ever be two implementations of this class (PosixEnv and WinEnv).  There is still a CustomEnvWrapper class that takes an Env and a FileSystem and wraps the Env calls with the input Env but uses the FileSystem for the FileSystem calls
-> Eliminates the public uses of the LegacyFileSystemWrapper.

With this change in place, there are effectively the following patterns of Env:
- "Base Env classes" (PosixEnv, WinEnv).  These classes implement the core Env functions (e.g. Threads) and have a hard-coded input FileSystem.  These classes inherit from CompositeEnv, implement the core Env functions (threads) and delegate the FileSystem-like calls to the input file system.
- Wrapped Composite Env classes (MemEnv).  These classes take in an Env and a FileSystem.  The core env functions are re-directed to the wrapped env.  The file system calls are redirected to the input file system
- Legacy Wrapped Env classes.  These classes take in an Env input (but no FileSystem).  The core env functions are re-directed to the wrapped env.  A "Legacy File System" is created using this env and the file system calls directed to the env itself.

With these changes in place, the PosixEnv becomes a singleton -- there is only ever one created.  Any other use of the PosixEnv is via another wrapped env.  This cleans up some of the issues with the env construction and destruction.

Additionally, there were places in the code that required had an Env when they required a FileSystem.  Many of these places would wrap the Env with a LegacyFileSystemWrapper instead of using the env->GetFileSystem().  These places were changed, thereby removing layers of additional redirection (LegacyFileSystem --> Env --> Env::FileSystem).

Pull Request resolved: https://github.com/facebook/rocksdb/pull/7703

Reviewed By: zhichao-cao

Differential Revision: D25762190

Pulled By: anand1976

fbshipit-source-id: 1a088e97fc916f28ac69c149cd1dcad0ab31704b
Signed-off-by: Changlong Chen <levisonchen@live.cn>

yhr committed:
Initial commmit for ZenFS

Merge experimental code to fix the long tail latency bug from westerndigitalcorporation/zenfs.
Since merge ZNS into terarkdb has been done earlier, only source codes are needed for now.
Other files will be removed in this patch for clean transportation.

The following is the original patch description of Hans:
----------------------------------------------------------------
ZenFS is a RocksDB plugin that allows RocksDB to store its files
on a raw zoned block device.

Signed-off-by: Changlong Chen <chenchanglong@bytedance.com>

zhichao-cao committed:
Handoff checksum Implementation (#7523)

Summary:
in PR https://github.com/facebook/rocksdb/issues/7419 , we introduce the new Append and PositionedAppend APIs to WritableFile at File System, which enable RocksDB to pass the data verification information (e.g., checksum of the data) to the lower layer. In this PR, we use the new API in WritableFileWriter, such that the file created via WritableFileWrite can pass the checksum to the storage layer. To control which types file should apply the checksum handoff, we add checksum_handoff_file_types to DBOptions. User can use this option to control which file types (Currently supported file tyes: kLogFile, kTableFile, kDescriptorFile.) should use the new Append and PositionedAppend APIs to handoff the verification information.

Pull Request resolved: https://github.com/facebook/rocksdb/pull/7523

Test Plan: add new unit test, pass make check/ make asan_check

Reviewed By: pdillinger

Differential Revision: D24313271

Pulled By: zhichao-cao

fbshipit-source-id: aafd69091ae85c3318e3e17cbb96fe7338da11d0
Signed-off-by: Changlong Chen <levisonchen@live.cn>

Commit:
zenfs: mv fs_zenfs -> env_zenfs

Resolve version differences, terarkdb(rocksdb 5.18.3) vs rocksdb 6.8.1

Because zenfs is implemented based on the higher version of rocksdb
coding, terarkdb is implemented based on the lower version of rocksdb
coding, this commit is to solve the version difference

Signed-off-by: Yuanliang Wang <wangyuanliang@bytedance.com>
Reviewed-by: Changlong Chen <chenchanglong@bytedance.com>

pdillinger committed:
Clarifying comments for Read() APIs (#8029)

Summary:
I recently discovered the confusing, undocumented semantics of
Read() functions in the FileSystem and Env APIs. I have added
clarification to the best of my reverse-engineered understanding, and
made a note in HISTORY.md for implementors to check their
implementations, as a subtly non-adherent implementation could lead to
RocksDB quietly ignoring some portion of a file.

Pull Request resolved: https://github.com/facebook/rocksdb/pull/8029

Test Plan: no code changes

Reviewed By: anand1976

Differential Revision: D26831698

Pulled By: pdillinger

fbshipit-source-id: 208f97ff6037bc13bb2ef360b987c2640c79bd03
Signed-off-by: Changlong Chen <levisonchen@live.cn>

yhr committed:
Make sure we have a single write-mount per file system

The following is the original patch description of Hans:
-----------------------------------------------------------------
Open the write FD with O_EXCL to make sure that there is a single
write-access user of the file system.

Also improve the error messages when opening the zoned block device.

Signed-off-by: Changlong Chen <chenchanglong@bytedance.com>

yhr committed:
Add a new command: df command to the zenfs utility

The following is the original patch description of Hans:
------------------------------------------------------------------
df shows the used and free capacity of the zenfs file system along
with space amplification.

Signed-off-by: Changlong Chen <chenchanglong@bytedance.com>

yhr committed:
Enable read-only mounting

The following is the original patch description of Hans:
------------------------------------------------------------
Allow read-only mounts, so a live instance can be opened for
checking disk space etc.

Signed-off-by: Changlong Chen <chenchanglong@bytedance.com>

yhr committed:
[PATCH 1/4] Add ZenFS, a RocksDB FileSystem for Zoned Block Devices

In order to support Zone Namespace SSD on TerarkDB, I transplanted
ZenFS from Hans Holmberg of WD and modified it, because ZenFS is
developed based on RocksDB v6.8.1, and TerarkDB is based on RocksDB
v5.18.3. Patch cannot be compiled directly. So I transplanted the
code but temporarily commented them out to prevent them from
participating in the compilation, and because the fs_posix.cc file
does not exist, I put some functions in the io_posix.cc file first.

The following is the original patch description of Hans:
----------------------------------------------------------------

Summary:

Host managed Zoned Block Devices enables an application to
do smart data placement by making informed decisions
on how to place data into the storage media's erase units.
This improves system write amplification and/or enables
greater media capacity utilization.

ZenFS is a simple file system that implements RockDBs
FileSystem interface to place files into zones on a raw
zoned block device. By separating files into zones
and utilizing the write life time hints to co-locate
data of simliar life times, the system write amplification
is greatly reduced while keeping the ZenFS capacity
overhead at a very resonable level.

ZenFS depends on libzbd to and Linux kernel 5.4 to do
zone management operations.

Some of the ideas and concepts in ZenFS are based on earlier
work done by Abutalib Aghayev and Marc Acosta.

Files are mapped into into a set of extents:

* Extents are block-aligned, continious regions on the block device
* Extents do not span across zones
* A zone may contain more than one extent
* Extents from different files may share zones

Log files and LOCK files are routed to a configurable
path in the default file system.

ZenFS is exceptionally lazy at current state of implementation
and does not do any garbage collection whatsoever. As files
gets deleted the used capacity zone counters drops and when
it reaches zero the zone can be reset and reused.

Metadata is stored in a rolling log in the first zones of the block device.

Each valid meta data zone contains:

* A superblock with the current sequence number and global filesytem metadata
* At least one snapshot of all files in the file system
* Update records for file deletions, renames, added extents etc.

Signed-off-by: Yuanliang Wang <wangyuanliang@bytedance.com>

yhr committed:
[PATCH 2/4] A add a file system management tool for ZenFS

TerarkDB supports Zone Namespace SSD porting ZenFS patch v2,
I removed the modification of the Makefile part, because
RocksDB v5.18.3 does not have this file, I will adapt this
part in the later patch.

The following is the original patch description of Hans:
----------------------------------------------------------------

Summary:

This adds a file system management tool for ZenFS. It is required
for setting up the metadata for the filesystem.

It also allows the user to list files on the file system and
can be extended in the future to do other management tasks
like offline garbage collection, backups etc.

Examples:

1. Create a zenfs file system on zoned block device /dev/nullb1 with
auxiliary file storage (logfiles, lockfiles) under /tmp/zenfs_nullb1 and
with a finishing threshold for zones of 5%. If the zone has less than
finish_threshold capacity left, no additional extents will be mapped
to the zone.

  ./zenfs mkfs --zbd=nullb1 --aux_path=/tmp/zenfs_nullb1 --finish_threshold=5
  ZenFS file system created. Free space: 31744 MB

2. List files:

  ./zenfs list --zbd=nullb1 --path=rocksdbtest/dbbench
  LOG
  LOCK
  000003.log
  CURRENT
  IDENTITY
  MANIFEST-000001
  OPTIONS-000005

Signed-off-by: Yuanliang Wang <wangyuanliang@bytedance.com>

yhr committed:
[PATCH 4/4] [NOT FOR UPSTREAM] Add zenfs branch info to README.md

TerarkDB supports Zone Namespace SSD porting ZenFS patch v4,
this part only does text matching

Signed-off-by: Yuanliang Wang <wangyuanliang@bytedance.com>

Commit:
zenfs: IOStatus -> Status [terarkdb vs rocksdb]

As ZenFS is based on RocksDB v6.8.3 and TerarkDB is based on v5.18.3,
the next version difference adaptation is required. The following series
of patches are to solve this problem.

Signed-off-by: Yuanliang Wang <wangyuanliang@bytedance.com>
Reviewed-by: Changlong Chen <chenchanglong@bytedance.com>

Commit:
zenfs: rm IOOptions IODebugContext [terarkdb vs rocksdb]

Signed-off-by: Yuanliang Wang <wangyuanliang@bytedance.com>
Reviewed-by: Changlong Chen <chenchanglong@bytedance.com>

Commit:
zenfs: Rename the base namespace [terarkdb vs rocksdb]

Signed-off-by: Yuanliang Wang <wangyuanliang@bytedance.com>
Reviewed-by: Changlong Chen <chenchanglong@bytedance.com>

Commit:
zenfs: Remove FileSystem class and related operations [terarkdb vs rocksdb]

Signed-off-by: Yuanliang Wang <wangyuanliang@bytedance.com>
Reviewed-by: Changlong Chen <chenchanglong@bytedance.com>

Commit:
zenfs: Some function adaptation, invalid function [terarkdb vs rocksdb]

Signed-off-by: Yuanliang Wang <wangyuanliang@bytedance.com>
Reviewed-by: Changlong Chen <chenchanglong@bytedance.com>

Commit:
zenfs: Expose the zenEnv interface publicly [terarkdb vs rocksdb]

Signed-off-by: Yuanliang Wang <wangyuanliang@bytedance.com>
Reviewed-by: Changlong Chen <chenchanglong@bytedance.com>

Commit:
zenfs: Add zenfs to compile [terarkdb vs rocksdb]

Now zenfs is integrated to provide functions

Signed-off-by: Yuanliang Wang <wangyuanliang@bytedance.com>
Reviewed-by: Changlong Chen <chenchanglong@bytedance.com>

Commit:
Fixed: Script minor issue

Signed-off-by: Yuanliang Wang <wangyuanliang@bytedance.com>
Reviewed-by: Changlong Chen <chenchanglong@bytedance.com>

Commit:
db_bench: Adapt fs_uri for ZenFS [terarkdb vs rocksdb]

Signed-off-by: Yuanliang Wang <wangyuanliang@bytedance.com>
Reviewed-by: Changlong Chen <chenchanglong@bytedance.com>

Commit:
zenfs tools: Add compilation support for ZenFS tools [terarkdb vs rocksdb]

Because TerakDB does not have a Makefile file, it uniformly
adapts to CMakefile.txt

Signed-off-by: Yuanliang Wang <wangyuanliang@bytedance.com>
Signed-off-by: chenchanglong <chenchanglong@bytedance.com>

Commit:
ZenFS: Fixed Segmentation fault (core dumped) triggered

ZenEnv file system created. Free space: 3596 MB
Initializing RocksDB Options from the specified file
Initializing RocksDB Options from command-line flags
RocksDB:    version 5.18
Date:       Wed Mar 10 09:28:45 2021
CPU:        64 * Intel(R) Xeon(R) Platinum 8260 CPU @ 2.40GHz
CPUCache:   16384 KB
Keys:       16 bytes each
Values:     800 bytes each (400 bytes after compression)
Entries:    100000
Prefix:    0 bytes
Keys per prefix:    0
RawSize:    77.8 MB (estimated)
FileSize:   39.7 MB (estimated)
Write rate: 0 bytes/second
Read rate: 0 ops/second
Compression: Snappy
Memtablerep: skip_list
Perf Level: 1
------------------------------------------------
Initializing RocksDB Options from the specified file
Initializing RocksDB Options from command-line flags
DB path: [db_test]
fillrandom   :       5.592 micros/op 178838 ops/sec;  139.2 MB/s
DB path: [db_test]
Segmentation fault (core dumped)

Signed-off-by: Changlong Chen <chenchanglong@bytedance.com>
Signed-off-by: Yuanliang Wang <wangyuanliang@bytedance.com>

Commit:
zenfs tools: zenfs tools adaptation support [terarkdb vs rocksdb]

Signed-off-by: Yuanliang Wang <wangyuanliang@bytedance.com>
Reviewed-by: Changlong Chen <chenchanglong@bytedance.com>

yhr committed:
EXPERIMENTAL: Add support for active/open zone limiting

PATCH 07/12 and 08/12 don't contain any source file changes, skip them.

The following is the original patch description of Hans:
----------------------------------------------------------------
Currently experimental

Signed-off-by: Changlong Chen <chenchanglong@bytedance.com>

yhr committed:
Dissallow deletes of files open for writing

PATCH 05/12 doesn't contain any source file changes, skip it.

The following is the original patch description of Hans:
----------------------------------------------------------------
We don't support deleting files open for writing,
so return an error if rocksdb tries to do this.

Signed-off-by: Changlong Chen <chenchanglong@bytedance.com>

yhr committed:
EXPERIMENTAL: Add script for splitting a device in half using dm-linear

Script for 2 instances on the single drive.

The following is the original patch description of Hans:
----------------------------------------------------------------
This allows two instances of zenfs to be used on a single drive.
Make sure the max/open limits are split as well.

Example:

MAX_OPEN=$((<max open on device> / 2))

blkzone reset /dev/nvme2n1
./split_dev.sh nvme2n1
./zenfs mkfs --aux-path=/root/test_aux-one --force --zbd=mapper/nvme2n1-ONE --finish_threshold=100 --max_active_zones=$MAX_OPEN --max_open_zones=$MAX_OPEN
./zenfs mkfs --aux-path=/root/test_aux-two --force --zbd=mapper/nvme2n1-TWO --finish_threshold=100 --max_active_zones=$MAX_OPEN --max_open_zones=$MAX_OPEN

Signed-off-by: Hans Holmberg <hans.holmberg@wdc.com>

yhr committed:
EXPERIMENTAL: Experimental allocator improvement

A substitutional plan for now.

The following is the original patch description of Hans:
----------------------------------------------------------------
Change the zone allocator to prioritize filling up zones.
This leads to higher space amplification (~2x) but reduces
write long-tail latencies greatly by reducing zone-finish
operations.

The better solution to the tail latency problems with
finish is to finish zones out of the allocator loop as
part of a background job, so that is what we should do in stead.

Signed-off-by: Changlong Chen <chenchanglong@bytedance.com>

mrambacher committed:
Fix spelling in comments in include/rocksdb/ (#8120)

Summary:
Ran a spell check over the comments in the include/rocksdb directory and fixed any mis-spellings.

There are still some variable names that are spelled incorrectly (like SizeApproximationOptions::include_memtabtles, SstFileMetaData::oldest_ancester_time) that were not fixed, as those would break compilation.

Pull Request resolved: https://github.com/facebook/rocksdb/pull/8120

Reviewed By: zhichao-cao

Differential Revision: D27366034

Pulled By: mrambacher

fbshipit-source-id: 6a3f3674890bb6acc751e9c5887a8fbb6adca5df
Signed-off-by: Changlong Chen <levisonchen@live.cn>

yhr committed:
EXPERIMENTAL: Support for async writes for buffered files

Using libaio for buffered writes in order to fix tail latency problem.
Remove makefile changes, since libaio is already one of TerarkDB's dependency.

The following is the original patch description of Hans:
----------------------------------------------------------------
This adds support for QD=2 writes for buffered (e.g WAL) writes.
This improves write tail latencies quite a bit and brings
us close to the tail latencies of regular file system writes.

Currently experimental, needs more work and validation before merging
to master.

This change also adds a dependency to libaio, which is not really needed,
we could work with the kernel ABI directly.

Signed-off-by: Changlong Chen <chenchanglong@bytedance.com>

Commit:
Add request_id in IODebugContext. (#8045)

Summary:
Add request_id in IODebugContext which will be populated by
    underlying FileSystem for IOTracing purposes. Update IOTracer to trace
    request_id in the tracing records. Provided API
    IODebugContext::SetRequestId which will set the request_id and enable
    tracing for request_id. The API hides the implementation and underlying
    file system needs to call this API directly.

Update DB::StartIOTrace API and remove redundant Env* from the
    argument as its not used and DB already has Env that is passed down to
    IOTracer.

Pull Request resolved: https://github.com/facebook/rocksdb/pull/8045

Test Plan: Update unit test.

Differential Revision: D26899871

Pulled By: akankshamahajan15

fbshipit-source-id: 56adef52ee5af0fb3060b607c3af1ec01635fa2b
Signed-off-by: Changlong Chen <levisonchen@live.cn>

Commit:
Integrated BlobDB for backup/restore support (#8129)

Summary:
Add support for blob files for backup/restore like table files.
    Since DB session ID is currently not supported for blob files (there is no place to store it in
    the header), so for blob files uses the
    kLegacyCrc32cAndFileSize naming scheme even if
    share_files_with_checksum_naming is set to kUseDbSessionId.

Pull Request resolved: https://github.com/facebook/rocksdb/pull/8129

Test Plan: Add new test units

Reviewed By: ltamasi

Differential Revision: D27408510

Pulled By: akankshamahajan15

fbshipit-source-id: b27434d189a639ef3e6ad165c61a143a2daaf06e
Signed-off-by: Changlong Chen <levisonchen@live.cn>

Commit:
Integrated BlobDB for backup/restore support (#8129)

Summary:
Add support for blob files for backup/restore like table files.
    Since DB session ID is currently not supported for blob files (there is no place to store it in
    the header), so for blob files uses the
    kLegacyCrc32cAndFileSize naming scheme even if
    share_files_with_checksum_naming is set to kUseDbSessionId.

Pull Request resolved: https://github.com/facebook/rocksdb/pull/8129

Test Plan: Add new test units

Reviewed By: ltamasi

Differential Revision: D27408510

Pulled By: akankshamahajan15

fbshipit-source-id: b27434d189a639ef3e6ad165c61a143a2daaf06e
Signed-off-by: Changlong Chen <levisonchen@live.cn>

levichen94 committed:
Add a compile script namespace rocksdb for abase

levichen94 committed:
Merge fs/zbd_zenfs into env/zbd_zenfs.

Signed-off-by: Changlong Chen <chenchanglong@bytedance.com>

levichen94 committed:
Merge fs/io_zenfs into env/io_zenfs.

Signed-off-by: Changlong Chen <chenchanglong@bytedance.com>

levichen94 committed:
Merge fs/fs_zenfs into env/env_zenfs.

Signed-off-by: Changlong Chen <chenchanglong@bytedance.com>

levichen94 committed:
Merge util/zenfs.cc into tools/zenfs_tool.cc

Signed-off-by: Changlong Chen <chenchanglong@bytedance.com>

levichen94 committed:
Adapt changes to terarkdb && Fix typo.

Signed-off-by: Changlong Chen <chenchanglong@bytedance.com>

levichen94 committed:
Remove "dev" prefix in earlier version

Signed-off-by: Changlong Chen <chenchanglong@bytedance.com>

levichen94 committed:
1. Do not treat deleting a missing file as an error anymore.
2. Add fs_uri to default bench script.

Signed-off-by: Changlong Chen <chenchanglong@bytedance.com>

Commit:
Revert miss changed buffer size.

mm304321141 committed:
Fix TerarkZipTableFactory::SanitizeOptions

mm304321141 committed:
Fix some warnings

Commit:
Change zenfs into submodule, fix cmakelists

Commit:
Merge in some headers and move zenfs to plugin

Commit:
Fix compile, merge in headers.

Commit:
WIP: Fix compile 1

Commit:
WIP: Fix compile 2, add ZenfsEnv to adopt fs

Commit:
WIP: Fix compile 3, All passed except zenfs exec with 'too many levels of symbolic links'

Commit:
Fix compile and warning, ready for test

royguo commented:
Please well-document these in the PR description:
1. Problem Addressing
2. Solution Details

levichen94 committed:
Fix zenfs dependency and remove default make install.

Commit:
Little change of API and remove useless code.

levichen94 committed:
change bench.py

levichen94 committed:
Fix compile and misc for dbbench

Commit:
New branch for bytendb test, get disk info from zns ssd

Commit:
Add API for zns disk space info, and remove useless code

Commit:
Remove protective marco inside env.h.

Commit:
--amend

Commit:
Backward compatible for cmake 'add_compile_definitions'

levichen94 committed:
Merge experimental branch for zenfs into bytendb test

levichen94 committed:
Enable bytedance metrics for terarkdb

levichen94 committed:
Bugfix : Change definition of used space size.

levichen94 committed:
Bugfix : Change definition of used space size.

levichen94 committed:
Update zenfs module.

skyzh committed:
add garbage collection for ZenFS

This PR adds GC for ZNS. Here's how it works:

* We use the statistics interface added in
  https://github.com/bzbd/zenfs/pull/6, where we can get what size
  each file occupies in each zone.
* We spawn a ZNSGC task in the background, which works like TTLGC.
  It polls statistics from ZenFS every 10s.
* The task will find all full zones with 50% garbage (by default)
  and mark files in that zone for GC.

Along with ZNS GC, other features are added.

* We added `zenfs_gc_ratio` option in db_bench command.
* We fixed a bug where files may be deleted before closed.

Signed-off-by: Alex Chi <iskyzh@gmail.com>

skyzh committed:
fix gc condition

Signed-off-by: Alex Chi <iskyzh@gmail.com>

skyzh committed:
support cf paths and add usage info on zenfs_gc_ratio

Signed-off-by: Alex Chi <iskyzh@gmail.com>

skyzh committed:
support multiple cf paths

Signed-off-by: Alex Chi <iskyzh@gmail.com>

skyzh committed:
dedup db paths

Signed-off-by: Alex Chi <iskyzh@gmail.com>

skyzh committed:
bump zenfs dependency and use assign in string

Signed-off-by: Alex Chi <iskyzh@gmail.com>

skyzh committed:
move ZoneStat to impl

Signed-off-by: Alex Chi <iskyzh@gmail.com>

skyzh committed:
add option to settings and do minor improvements

Signed-off-by: Alex Chi <iskyzh@gmail.com>

skyzh committed:
remove debug message

Signed-off-by: Alex Chi <iskyzh@gmail.com>

skyzh committed:
use 0.25 as default db_bench zenfs_gc_ratio

Signed-off-by: Alex Chi <iskyzh@gmail.com>

mm304321141 committed:
Fix DBImpl::PurgeObsoleteFiles null ptr crash

levichen94 committed:
Update space info & fix delete file double free bug for zenfs.

skyzh committed:
add ZenFS tracing facilities

Signed-off-by: Alex Chi <iskyzh@gmail.com>

skyzh committed:
fix negative lifetime hint

skyzh committed:
tracing: use global metrics reporter (#97)

Signed-off-by: Alex Chi <iskyzh@gmail.com>

levichen94 committed:
Close logger when write buffered log, update zenfs for more metrics

levichen94 committed:
Close logwriter at enqueue delete scheduler.

levichen94 committed:
Bugfix : logger double free.

levichen94 committed:
Close old logger when creating new one.

levichen94 committed:
Add active log number printing.

levichen94 committed:
WIP: debug log close coredump

levichen94 committed:
Bugfix: Close logwriter for limited active zones.

Commit:
fix segfault issue when running db_bench dafult parameters

Commit:
add constructor to LatencyHistGuard to support split wal and sst in panel

yapple committed:
Enhancement metric report

* Merge enhancement metric repor with LatencyHistLoggedGuard from
  upstream
* Adoption for zenfs

levichen94 committed:
Update gitignore

levichen94 committed:
Default constructor for latencyhistguard.

levichen94 committed:
Debug: Using LoggedGuard in write impl
